How we can generate contextualized word embeddings

In [1]:
from transformers import AutoModel, AutoTokenizer

C:\Users\João Alves\AppData\Roaming\Python\Python311\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [2]:
# Load a tokenizer
tokenizer = AutoTokenizer.from_pretrained("microsoft/deberta-base")

# Load a language model
model = AutoModel.from_pretrained("microsoft/deberta-v3-base")

# Tokenize the sentence
tokens = tokenizer("Hello world", return_tensors="pt")

# Process the tokens
output = model(**tokens)[0]

C:\Users\João Alves\AppData\Roaming\Python\Python311\site-packages\huggingface_hub\file_download.py:143: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\João Alves\.cache\huggingface\hub\models--microsoft--deberta-base. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)
C:\Users\João Alves\AppData\Roaming\Python\Python311\site-packages\huggingface_hub\file_down

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


This code downloads a pretrained tokenizer and model, then uses them to process the string "Hello world".

The output of the model is then saved in the output variable.

In [3]:
output.shape

torch.Size([1, 4, 768])

- Four tokens, each one embedded in a vector of 768 values
- The first dimension is the batch dimension used in cases when we want to send multiple input sentences to the model at the same time.

What are those 4 tokens?

In [4]:
for token in tokens['input_ids'][0]:
    print(tokenizer.decode(token))

[CLS]
Hello
 world
[SEP]


In [6]:
# Raw output of the language model
output

tensor([[[ 0.1778,  0.3151, -0.0893,  ..., -0.0230,  0.1853,  0.0352],
         [ 1.3140,  0.3352,  0.3421,  ..., -0.4859, -0.0331, -0.3244],
         [ 0.7811,  1.0871,  0.2801,  ..., -0.1956, -0.0848,  0.1967],
         [ 0.2305,  0.3857,  0.0139,  ...,  0.0109,  0.2342, -0.0375]]],
       grad_fn=<NativeLayerNormBackward0>)